### PR TITLE
util,executor: add a package and estimate memory used by hash join

### PR DESCRIPTION
**Pull request by tiancaiamao**
It's not easy to know precisely how much memory used by a session, but we can do some estimation.
This PR is towards the way to improve the situation.

### CODE REVIEW

**Comment by shenli** [comment]
You can try this branch to see how accurately we could get the memory usage.

**Comment by tiancaiamao** [comment]
PTAL @zimulala @shenli @coocood @hanfei1991

File: context/context.go [hunk 1]
```diff
@@ -78,4 +80,6 @@ const (
 	QueryString basicCtxType = 1
 	// Initing is the key for indicating if the server is running bootstrap or upgrad job.
 	Initing basicCtxType = 2
+	// MemoryStatistics is the key for Memory allocation statistics.
```
Comment (andelf) [inline context/context.go]: lowercase Memory

File: executor/join.go [hunk 2]
```diff
@@ -352,9 +368,13 @@ func (e *HashJoinExec) runJoinWorker(idx int) {
 			}
 			if len(result.rows) >= maxRowsCnt {
 				e.resultCh <- result
-				result = &execResult{rows: make([]*Row, 0, maxRowsCnt)}
+				result = &execResult{
+					rows:     make([]*Row, 0, maxRowsCnt),
+					memUsage: e.memUsage.Open(),
```
Comment (coocood) [inline executor/join.go]: One record per big row is too expensive.

Comment (tiancaiamao) [inline executor/join.go]: So, any better suggestion?

**Comment by tiancaiamao** [comment]
Hard to do it precisely and introduce much complexity.

**Pull request closed without merging**